internal/refine/refine.go

```diff
@@ -0,0 +1,121 @@
+package refine
+
+import (
+	"errors"
+
+	"github.com/eigerco/strawberry/internal/block"
+	"github.com/eigerco/strawberry/internal/crypto"
+	"github.com/eigerco/strawberry/internal/polkavm"
+	"github.com/eigerco/strawberry/internal/polkavm/host_call"
+	"github.com/eigerco/strawberry/internal/polkavm/interpreter"
+	"github.com/eigerco/strawberry/internal/state"
+	"github.com/eigerco/strawberry/internal/work"
+	"github.com/eigerco/strawberry/pkg/serialization/codec/jam"
+)
+
+const (
+	RefineCost = 10
+)
+
+var (
+	ErrBad = errors.New("service’s code was not available for lookup") // BAD
+	ErrBig = errors.New("code beyond the maximum size allowed")        // BIG
+)
+
+type Refine struct {
+	state state.State
+}
+
+// InvokePVM ΨR(H, NG, NS , H, Y, X, H, Y, ⟦G⟧, ⟦Y⟧, N) → (Y ∪ J, ⟦Y⟧)
+func (r *Refine) InvokePVM(
+	serviceCodePredictionHash crypto.Hash, gas uint64, serviceIndex block.ServiceId,
+	workPackageHash crypto.Hash, workPayload []byte, refinementContext block.RefinementContext,
+	authorizerHash crypto.Hash, authorizerHashOutput []byte, importedSegments []polkavm.Segment,
+	extrinsicDataBlobs [][]byte, exportOffset uint64,
```

Comment:
Consider passing the `Item` and `Package` structs from the `work` package to reduce the parameters.